plugins/events-backend-module-http/src/service/types.ts

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2022 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Request } from 'express';
+
+/**
+ * Details for how to respond to the rejection
+ * of the received HTTP request transmitting an event payload.
+ *
+ * @public
+ */
+export interface EventHttpRequestRejection {
+  status?: number;
+  payload?: any;
+}
+
+/**
+ * Validator used to check the received HTTP request
+ * transmitting an event payload.
+ *
+ * E.g., it can be used for signature verification like
+ * for GitHub webhook events
+ * (https://docs.github.com/en/developers/webhooks-and-events/webhooks/creating-webhooks#secret)
+ * or other kinds of checks.
+ *
+ * Returns rejection details with options to control the response.
+ *
+ * @public
+ */
+export type EventHttpRequestValidator = (
+  request: Request,
```

Comment:
> whether the payload is already read and parsed, on the request object

I was thinking about it, but didn't see any issues with reading/calling `request.body` multiple times. Maybe my lack of knowledge about this lib. The body is already parsed at `request.body` as far as I understood.

So essentially, it is just passing the request object to reach validator and calls `request.body` itself later.

Reply:
> How would you like to have a second argument instead of returning an outcome?

I don't really fully get what you mean here.

> Otherwise, it seems like there's a setup for a discriminated union in the return type. Then maybe there should be a `type: 'reject'` explicitly on `EventHttpRequestRejection`

Same here.

At the moment, you can only pass a single validator (however, I though about adding a compound one).

The validator itself has the task to either accept or reject the request depending on its own logic. If it accepts the request, we move on and process it "normally" -- we publish the event payload and return with status 202.

If the validator rejects the request, we want stop and return an error state and error payload. As there might be different rejection reasons, `EventHttpRequestRejection` allows to customize the response status code and payload with a default (403 + `{}`).

In the first iteration, I've returned a boolean, but that was not enough. I decided to return `undefined` for the success case.
Probably, what you see as problematic, however I may misinterpret that.

Looking at your code example, `context` could be used to set the rejection (with optional response customization) instead of returning anything. 🤔

Reply:
I gave it a try. Let me know what you think about the new version.